macaroonbakery/oven.py

```diff
@@ -0,0 +1,239 @@
+
+        def validator(condition):
+            conditions.append(condition)
+            return True
+        v.satisfy_general(validator)
+        v.verify(macaroons[0], root_key, macaroons[1:])
+        if (self.ops_store is not None
+            and len(ops) == 1
+                and ops[0].entity.startswith("multi-")):
+            # It's a multi-op entity, so retrieve the actual operations
+            # it's associated with.
+            ops = self.ops_store.get_opts(ops[0].entity)
+            if ops is None:
+                raise KeyError(
+                    "cannot get operations for {}".format(ops[0].entity))
+
+        return ops, conditions
+
+
+def _decode_macaroon_id(id):
+    storage_id = b''
+    base64_decoded = False
+    first = id[:1]
+    if first == b'A':
+        # The first byte is not a version number and it's 'A', which is the
+        # base64 encoding of the top 6 bits (all zero) of the version number 2
+        # or 3, so we assume that it's the base64 encoding of a new-style
+        # macaroon id, so we base64 decode it.
+        #
+        # Note that old-style ids always start with an ASCII character >= 4
+        # (> 32 in fact) so this logic won't be triggered for those.
+        try:
+            dec = utils.raw_b64decode(id.decode('utf-8'))
+            # Set the id only on success.
+            id = dec
+            base64_decoded = True
+        except:
+            # if it's a bad encoding, we'll get an error which is fine
+            pass
+
+    # Trim any extraneous information from the id before retrieving
+    # it from storage, including the UUID that's added when
+    # creating macaroons to make all macaroons unique even if
+    # they're using the same root key.
+    first = six.byte2int(id[:1])
+    if first == bakery.BAKERY_V2:
+        # Skip the UUID at the start of the id.
+        storage_id = id[1 + 16:]
+    if first == bakery.BAKERY_V3:
+        id1 = id_pb2.MacaroonId.FromString(id[1:])
+        if len(id1.ops) == 0 and len(id1.ops[0].actions) == 0:
+            raise ValueError("no operations found in macaroon")
+
+        ops = []
+        for op in id1.ops:
+            for action in op.actions:
+                ops.append(checker.Op(op.entity, action))
+        return id1.storageId, ops
+
+    if not base64_decoded and _is_lower_case_hex_char(first):
+        # It's an old-style id, probably with a hyphenated UUID.
+        # so trim that off.
+        last = id.rfind(b'-')
+        if last >= 0:
+            storage_id = id[0:last]
+    return storage_id, [checker.LOGIN_OP]
+
+
+def _is_lower_case_hex_char(b):
+    if ord('0') <= b <= ord('9'):
+        return True
+    if ord('a') <= b <= ord('f'):
+        return True
+    return False
+
+
+def canonical_ops(ops):
+    ''' Returns the given operations array sorted with duplicates removed.
+
+    @param ops checker.Ops
+    @return: checker.Ops
+    '''
+
+    new_ops = sorted(set(ops), key=lambda x: (x.entity, x.action))
```

Comment:
Shouldn't this be returning an array of Op instances?

Reply:
this is what it does, it returns an array of ops sorted by entity then action.

Reply:
Ah! Cool, I hadn't understood the role of "key".